Codegen UPDATE

seojungmin opened the pull request:
This PR is for processing UPDATE within Codegen. The initial code has come from #656, and much of the code has been refactored and modified for a clean PR, e.g. the global runtime states introduced in the previous PR are all removed and localized, and a helper class like that in #687 has been introduced to accommodate many of the changes.  

`PerformUpdate()` that contains the original executor logic has been touched for a clear code representation.  Although it is more readable now, it is not still in its best condition.  There is also serialization/deserialization for the target/direct_map lists while being passed through the translator, and this may need some attention to be improved near future.

In addition, a bug on handling VARCHAR in codegen has been fixed, which is not related to UPDATE.   The original peloton value system has been including null terminators in its VARCHAR's length value, but that of codegen has been changed to behave in the same way.

seojungmin commented:
The Coveralls program seemed to behave a little bit abnormally.  It reported that the coverage went down about 11%, which was unlikely considering the change that had been made in this PR.  @pmenon helped me to rerun the travis, and it seems to report the coverage correctly.

seojungmin commented:
I will rebase this, now.

seojungmin commented:
This PR can be merged. Review comments are always welcome.

pmenon commented:
@seojungmin Can you provide high-level points on how this works? Also add comments in the code describing each component.

seojungmin commented:
@pmenon Here are some high-level points. I added comments on the last commit.  

This PR introduces two new modules, and also changes roles of some previous modules:

[New]
- _Update translator_ builds the information required for update and hand it over to the updater.
- _Updater_ is a helper for the update execution, and deals with updating with and without primary key.  It also calls TransactionRuntime to find out(or to acquire) the ownership of the tuple. Depending on the ownership it has different execution strategies.  The execution strategy is the same as before, but the code is clean up.

[Role Changes]
- _TransactionRuntime_ used to do all the execution work, but I moved these roles to Deleter and I only left helper APIs for acquiring ownership.
- _Deleter_ used to be only a bridge to TransactionRuntime, but it now has all the execution logic inside, which is consistent with Updater.

Since he update execution logic was so huge and complex, these separations and changes were inevitable.   After all, the translators do whatever codegen can do, and pass over to Updater/Inserter/Deleter which actually do the execution jobs that cannot be handled by codegen.

pmenon commented on src/codegen/operator/update_translator.cpp:
```diff
@@ -0,0 +1,157 @@
+  auto *scan =
+      static_cast<const planner::AbstractScan *>(update_plan_.GetChild(0));
+  std::vector<const planner::AttributeInfo *> ais;
+  scan->GetAttributes(ais);
```
Shouldn't this be done in `UpdatePlan`? At this point, the plans should already know all attribute information.

pmenon commented on src/codegen/updater.cpp:
```diff
@@ -0,0 +1,173 @@
+namespace peloton {
+namespace codegen {
+
+void Updater::Init(concurrency::Transaction *txn, storage::DataTable *table,
```
Modify the initialization by removing `Transaction`, and add `ExecutorContext`. This way, you get both the memory pool and transaction handle at initialization time. You can then remove `GetPool()` and remove `ExecutorContext` from `Update()`

pmenon commented on src/codegen/updater.cpp:
```diff
@@ -0,0 +1,173 @@
+  table_ = table;
+  // Target list is kept since it is required at a new version update
+  target_list_.reset(
+      new TargetList(target_vector, target_vector + target_vector_size));
```
Does `target_list_` have to be a unique_ptr?

pmenon commented on src/include/codegen/operator/update_translator.h:
```diff
@@ -0,0 +1,73 @@
+ private:
+  void SetTargetValue(llvm::Value *target_val_vec, llvm::Value *target_id,
+                      type::Type type, llvm::Value *value,
+                      llvm::Value *length) const;
```
Not sure this is implemented.

pmenon commented on src/include/codegen/operator/update_translator.h:
```diff
@@ -0,0 +1,73 @@
+  storage::DataTable *target_table_;
+
+  // Boolean value representing whether it is a primary key update or not
+  bool update_primary_key_;
```
Not sure if this is used.

pmenon commented on src/include/codegen/transaction_runtime.h:
```diff
@@ -13,6 +13,8 @@
 
 #include <cstdint>
+#include "type/types.h"
+#include "common/item_pointer.h"
```
Is this needed?
